server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java

```diff
@@ -172,53 +172,60 @@ private static int getTotalNewShards(Index index, ClusterState currentState, int
         // against such mixed nodes for production use anyway.
         int frozenNodeCount = nodeCount(state, ShardLimitValidator::hasFrozen);
         int normalNodeCount = nodeCount(state, ShardLimitValidator::hasNonFrozen);
-        return checkShardLimit(newShards, state, getShardLimitPerNode(), normalNodeCount, NORMAL_GROUP).or(
-            () -> checkShardLimit(newFrozenShards, state, shardLimitPerNodeFrozen.get(), frozenNodeCount, "frozen")
-        );
+
+        var result = checkShardLimit(newShards, state, getShardLimitPerNode(), normalNodeCount, NORMAL_GROUP);
+        // fail-fast: in case there's no room on the `normal` nodes, just return the result of that check.
+        if (result.canAddShards() == false) {
+            return result;
+        }
+        return checkShardLimit(newFrozenShards, state, shardLimitPerNodeFrozen.get(), frozenNodeCount, FROZEN_GROUP);
     }
 
     /**
-     * This method decides whether there is enough room in the cluster to add the given number of shards with the given number of replicas
-     * without exceeding the "cluster.max_shards_per_node" setting if the shards are going on normal nodes. This check does not guarantee
-     * that the number of shards can be added, just that there is theoretically room to add them without exceeding the shards per node
-     * configuration.
+     * This method checks whether there is enough room in the cluster to add the given number of shards with the given number of replicas
+     * without exceeding the "cluster.max_shards_per_node" setting for _normal_ nodes. This check does not guarantee that the number of
+     * shards can be added, just that there is theoretically room to add them without exceeding the shards per node configuration.
      * @param numberOfNewShards The number of primary shards that we want to be able to add to the cluster
      * @param replicas          The number of replicas of the primary shards that we want to be able to add to the cluster
      * @param state             The cluster state, used to get cluster settings and to get the number of open shards already in the cluster
-     * @return True if there is room to add the requested number of shards to the cluster, and false if there is not
      */
-    public static boolean canAddShardsToCluster(int numberOfNewShards, int replicas, ClusterState state) {
-        var clusterSettings = state.getMetadata().settings();
-        var maxShardsPerNode = SETTING_CLUSTER_MAX_SHARDS_PER_NODE.get(clusterSettings);
-        var nodeCount = nodeCount(state, ShardLimitValidator::hasNonFrozen);
-        var errorMessage = checkShardLimit(numberOfNewShards * (1 + replicas), state, maxShardsPerNode, nodeCount, NORMAL_GROUP);
-        return errorMessage.isEmpty();
+    public static Result checkShardLimitForNormalNode(int numberOfNewShards, int replicas, ClusterState state) {
```

Comment:
A couple of minor things -- (1) this probably ought to be `checkShardLimitForNormalNodes` (plural), right? (2) You removed the @return javadoc (although it's pretty obvious what's returned here).

Reply:
1. you're right. will update with the plural version.
2. that's why I removed it. Right now the `Result` object has its own javadoc, so wanted to avoid an obvious comment